category_theory/eq_to_hom.lean

```diff
@@ -0,0 +1,78 @@
+-- Copyright (c) 2018 Reid Barton. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Reid Barton, Scott Morrison
+
+import category_theory.isomorphism
+import category_theory.functor_category
+
+universes u v u' v'
+
+namespace category_theory
+
+variables {C : Type u} [𝒞 : category.{u v} C]
+include 𝒞
+
+def eq_to_hom {X Y : C} (p : X = Y) : X ⟶ Y := by rw p; exact 𝟙 _
+
+@[simp] lemma eq_to_hom_refl (X : C) (p : X = X) : eq_to_hom p = 𝟙 X := rfl
+@[simp] lemma eq_to_hom_trans {X Y Z : C} (p : X = Y) (q : Y = Z) :
+  eq_to_hom p ≫ eq_to_hom q = eq_to_hom (p.trans q) :=
+by cases p; cases q; simp
+@[simp] lemma eq_to_hom_trans_assoc {X Y Z W : C} (p : X = Y) (q : Y = Z) (f : Z ⟶ W) :
+  eq_to_hom p ≫ (eq_to_hom q ≫ f) = eq_to_hom (p.trans q) ≫ f :=
+by cases p; cases q; simp
+
+def eq_to_iso {X Y : C} (p : X = Y) : X ≅ Y :=
+⟨eq_to_hom p, eq_to_hom p.symm, by simp, by simp⟩
+
+@[simp] lemma eq_to_iso.hom {X Y : C} (p : X = Y) : (eq_to_iso p).hom = eq_to_hom p :=
+rfl
+
+@[simp] lemma eq_to_iso_refl (X : C) (p : X = X) : eq_to_iso p = iso.refl X := rfl
+@[simp] lemma eq_to_iso_trans {X Y Z : C} (p : X = Y) (q : Y = Z) :
+  eq_to_iso p ≪≫ eq_to_iso q = eq_to_iso (p.trans q) :=
+by ext; simp
+
+variables {D : Type u'} [𝒟 : category.{u' v'} D]
+include 𝒟
+
+namespace functor
+
+/-- Proving equality between functors. This isn't an extensionality lemma,
+  because usually you don't really want to do this. -/
+lemma ext {F G : C ⥤ D} (h_obj : ∀ X, F.obj X = G.obj X)
+  (h_map : ∀ X Y f, F.map f = eq_to_hom (h_obj X) ≫ G.map f ≫ eq_to_hom (h_obj Y).symm) :
```

Comment:
It seems like this could be stated more abstractly if you move the `eq_to_hom (h_obj Y).symm` to the LHS of the equation, because then it looks like a naturality condition. This is something like `eq_to_hom` is a nat_trans between `F` and `G`?